packages/components/src/components/entities/actions.ts

```diff
@@ -44,3 +53,232 @@ export function getSampleDeleteConfirmationData(selectionKey: string, rowIds?: A
+            query: schemaQuery.getQuery(),
+            value: data
+        })
+    ]);
+}
+
+function extractFromRow(row: Map<string, any>): IEntityTypeOption {
+    const name = row.getIn(['Name', 'value']);
+    return {
+        label: name,
+        lsid: row.getIn(['LSID', 'value']),
+        rowId: row.getIn(['RowId', 'value']),
+        value: name.toLowerCase(), // we match values on lower case because (at least) when parsed from an id they are lower case
+        query: name
+    }
+}
+
+function getChosenParentData(model: EntityIdCreationModel, parentSchemaQueries: Map<string, SchemaQuery>, allowParents: boolean) : Promise<Partial<EntityIdCreationModel>> {
+    return new Promise((resolve, reject) => {
+        let entityParents = Map<string, List<EntityParentType>>();
+        parentSchemaQueries.forEach((schemaQuery) => {
+            entityParents = entityParents.set(schemaQuery.queryName, List<EntityParentType>());
+        });
+        if (allowParents) {
+            const parentSchemaNames = parentSchemaQueries.keySeq();
+            initParents(model.originalParents, model.selectionKey).then(
+                (chosenParents) => {
+                    // if we have an initial parent, we want to start with a row in the grid (entityCount = 1) otherwise we start with none
+                    const parentRep = chosenParents.find((parent) => parent.value !== undefined && parentSchemaNames.contains(parent.schema));
+                    let validEntityCount = parentRep ? 1 : 0;
+
+                    if (validEntityCount === 1) {
+                        resolve({
+                            entityCount: validEntityCount,
+                            entityParents: entityParents.set(parentSchemaQueries.get(parentRep.schema).queryName, chosenParents),
+                        });
+                    }
+                    else { // if we did not find a valid parent, we clear out the parents and selection key from the model as they aren't relevant
+                        resolve({
+                            originalParents: undefined,
+                            selectionKey: undefined,
+                            entityParents,
+                            entityCount: 0
+                        })
+                    }
+                }
+            ).catch((reason) => {
+                console.error(reason);
+                reject(reason);
+            });
+        }
+        else {
+            resolve({
+                originalParents: undefined,
+                selectionKey: undefined,
+                entityParents,
+                entityCount: 0
+            })
+        }
+    })
+}
+
+// get back a map from the typeListQueryName (e.g., 'SampleSet') and the list of options for that query
+// where the schema field for those options is the typeSchemaName (e.g., 'samples')
+function getEntityTypeOptions(typeListSchemaQuery: SchemaQuery, typeSchemaName: string) : Promise<Map<string, List<any>>> {
```

Comment:
maybe add a todo item to filter dataclass by category. would need to think how to not make it not too hardcoded...